### ISSUE TITLE
Parallel optimization breaks results in a very bad way

### ISSUE THREAD

**Issue by davrot** [numba/numba#7170]
Hi,

I found an example that is broken by the parallel=true & prange optimization. (broken version: numba_4_debug.py ; workaround with correct results in numba_4.py). It is a nasty problem because there is no indication for the user that such a dramatic error occurred.

The results need to be -- by definition of the problem -- in the value range 0 .... 9999. 
numba_4_debug.py this is slightly out of range but it is order of magnitudes faster :-) :

Result vector: 
[                   0 14685055086129564619 14685055086129564619 ...
                    0                    0                    0]

Somehow the optimization interacts with the initialization of the Result vector 
From the debug information (full info below):
Result = numpy.zeros((NumberOfIterations), dtype=numpy.uint64)---------------------------| #0

In the workaround I take the Result vector outside of the function that is numba.njit-ed and then the results are correct.
 
best wishes
David

[numba_4_debug.py.txt](https://github.com/numba/numba/files/6740586/numba_4_debug.py.txt)
[numba_4.py.txt](https://github.com/numba/numba/files/6740587/numba_4.py.txt)

Tested on Python 3.9.5 and Numba 0.53.1

Debug info:
================================================================================
 Parallel Accelerator Optimizing:  Function main,
/home/davrot/Talk/numba_4_debug.py (22)
================================================================================

Parallel loop listing for  Function main, /home/davrot/Talk/numba_4_debug.py (22)
---------------------------------------------------------------------------------------------|loop #ID
@numba.njit(numba.types.uint64[:](numba.types.uint64, numba.types.uint64), parallel=True)    |
def main(NumberOfIterations, NumerOfNeurons):                                                |
    Result = numpy.zeros((NumberOfIterations), dtype=numpy.uint64)---------------------------| #0
                                                                                             |
    for i in numba.prange(NumberOfIterations):-----------------------------------------------| #3
        H = numpy.random.random(NumerOfNeurons).astype(numpy.float64)------------------------| #1
        H /= numpy.sum(H)--------------------------------------------------------------------| #2
        Temp = GetSpike(H, NumerOfNeurons)                                                   |
        Result[i] = Temp                                                                     |
    return Result                                                                            |
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
  Trying to fuse loops #0 and #3:
    - fusion succeeded: parallel for-loop #3 is fused into for-loop #0.
----------------------------- Before Optimisation ------------------------------
Parallel region 0:
+--0 (parallel)
   +--1 (parallel)
   +--2 (parallel)

--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel region 0:
+--0 (parallel)
   +--1 (serial)
   +--2 (serial)

Parallel region 0 (loop #0) had 0 loop(s) fused and 2 loop(s) serialized as part
 of the larger parallel loop (#0).
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

---------------------------Loop invariant code motion---------------------------
Allocation hoisting:
The memory allocation derived from the instruction at
/home/davrot/Talk/numba_4_debug.py (27) is hoisted out of the parallel loop
labelled #0 (it will be performed before the loop is executed and reused inside
the loop):
   Allocation:: H = numpy.random.random(NumerOfNeurons).astype(numpy.float64)
    - numpy.empty() is used for the allocation.

Instruction hoisting:
loop #0:
  Has the following hoisted:
    $expr_out_var.39 = const(uint64, 0)
    $32load_global.2 = global(numpy: <module 'numpy' from '/data_1/davrot/Py3.9/lib/python3.9/site-packages/numpy/__init__.py'>)
    $np_g_var.41 = global(np: <module 'numpy' from '/data_1/davrot/Py3.9/lib/python3.9/site-packages/numpy/__init__.py'>)
    $empty_attr_attr.42 = getattr(value=$np_g_var.41, attr=empty)
    $np_typ_var.43 = getattr(value=$np_g_var.41, attr=float64)
    $40call_method.6 = call $empty_attr_attr.42(NumerOfNeurons, $np_typ_var.43, func=$empty_attr_attr.42, args=[Var(NumerOfNeurons, numba_4_debug.py:24), Var($np_typ_var.43, numba_4_debug.py:27)], kws=(), vararg=None)
    $44load_global.8 = global(numpy: <module 'numpy' from '/data_1/davrot/Py3.9/lib/python3.9/site-packages/numpy/__init__.py'>)
    $46load_attr.9 = getattr(value=$44load_global.8, attr=float64)
    $66load_global.17 = global(GetSpike: CPUDispatcher(<function GetSpike at 0x7fd67b619670>))
  Failed to hoist the following:
    dependency: $36load_method.4 = getattr(value=$34load_attr.3, attr=random)
    dependency: $42load_method.7 = getattr(value=$40call_method.6, attr=astype)
    dependency: H = call $42load_method.7($push_getattr_to_block.227, func=$42load_method.7, args=[Var($push_getattr_to_block.227, numba_4_debug.py:27)], kws=(), vararg=None)
    dependency: H_shape.32 = getattr(value=H, attr=shape)
    dependency: H_size0.33 = static_getitem(value=H_shape.32, index=0, index_var=None, fn=<built-in function getitem>)
    dependency: Temp = call $push_global_to_block.229($H.1.56, NumerOfNeurons, func=$push_global_to_block.229, args=[Var($H.1.56, numba_4_debug.py:28), Var(NumerOfNeurons, numba_4_debug.py:24)], kws=(), vararg=None)
--------------------------------------------------------------------------------

**Comment by gmarkall** [comment]
Reproducer formatted for copy/paste:

```python
import numpy
import numba
import time

@numba.njit(numba.types.uint64(numba.types.float64[::1], numba.types.uint64))
def GetSpike(H, N_H):

        RandomNumber = numpy.float64(numpy.random.random())

        Summation = numpy.float64(0.0)

        for i in range(0, N_H - 1):
                Summation += H[i]

                if RandomNumber <= Summation:
                        Output = numpy.uint64(i)
                        return Output

        Output = numpy.uint64(N_H - 1)
        return Output

@numba.njit(numba.types.uint64[:](numba.types.uint64, numba.types.uint64), parallel=True)
def main(NumberOfIterations, NumerOfNeurons):
    Result = numpy.zeros((NumberOfIterations), dtype=numpy.uint64)

    for i in numba.prange(NumberOfIterations):
        H = numpy.random.random(NumerOfNeurons).astype(numpy.float64)
        H /= numpy.sum(H)
        Temp = GetSpike(H, NumerOfNeurons)
        Result[i] = Temp
    return Result

if __name__ == "__main__":
    NumberOfIterations = numpy.uint64(10000)
    NumerOfNeurons = 10000

    tic = time.perf_counter()
    Result = main(NumberOfIterations, NumerOfNeurons)
    toc = time.perf_counter()
    print(str(numpy.round(toc-tic,5)) + " sec")
    main.parallel_diagnostics(level=4)
    print(Result)
```

WAR formatted for copy/paste:

```python
import numpy
import numba
import time

@numba.njit(numba.types.uint64(numba.types.float64[::1], numba.types.uint64))
def GetSpike(H, N_H):

        RandomNumber = numpy.float64(numpy.random.random())

        Summation = numpy.float64(0.0)

        for i in range(0, N_H - 1):
                Summation += H[i]

                if RandomNumber <= Summation:
                        Output = numpy.uint64(i)
                        return Output

        Output = numpy.uint64(N_H - 1)
        return Output

@numba.njit(numba.types.uint64[:](numba.types.uint64[::1],numba.types.uint64, numba.types.uint64), parallel=True)
def main(Result, NumberOfIterations, NumerOfNeurons):

    for i in numba.prange(NumberOfIterations):
        H = numpy.random.random(NumerOfNeurons).astype(numpy.float64)
        H /= numpy.sum(H)
        Temp = GetSpike(H, NumerOfNeurons)
        Result[i] = Temp
    return Result

if __name__ == "__main__":
    NumberOfIterations = numpy.uint64(10000)
    NumerOfNeurons = 10000
    Result = numpy.zeros((NumberOfIterations), dtype=numpy.uint64)

    tic = time.perf_counter()
    Result = main(Result, NumberOfIterations, NumerOfNeurons)
    toc = time.perf_counter()
    print(str(numpy.round(toc-tic,5)) + " sec")
    print(Result)
```

**Comment by gmarkall** [comment]
This comment from the Parallel Accelerator Optimizing report seems suspicious:

```
---------------------------Loop invariant code motion---------------------------
Allocation hoisting:
The memory allocation derived from the instruction at repro.py (27) is hoisted 
out of the parallel loop labelled #0 (it will be performed before the loop is 
executed and reused inside the loop):
   Allocation:: H = numpy.random.random(NumerOfNeurons).astype(numpy.float64)
    - numpy.empty() is used for the allocation.
```

It seems like `H` was private to each thread before the hoisting, but is shared after it - is that the case? What do you think, @DrTodd13 ?

**Comment by stuartarchibald** [comment]
Smaller reproducer:
```python
import numpy
import numba

@numba.njit(parallel=True)
def main(x):
    res = numpy.zeros((x))
    for i in numba.prange(x):
        H = numpy.ones((1,), numpy.float64)
        H += numpy.sum(H)
        res[i] = 2
    return res

print(main(100))
```

**Comment by DrTodd13** [comment]
@davrot A fix I just did for issue #7105 seems to also fix this issue.

**Comment by stuartarchibald** [comment]
> @davrot A fix I just did for issue #7105 seems to also fix this issue.

Thanks for confirming @DrTodd13

**Comment by stuartarchibald** [comment]
Fixed by https://github.com/numba/numba/pull/7230. Closing.
